File: widget/src/main/java/ru/kest/trainswidget/converters/YandexToDomainConverter.java
package ru.kest.trainswidget.converters;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.List;

import ru.kest.trainswidget.model.domain.TrainThread;
import ru.kest.trainswidget.model.yandex.ScheduleResponse;

/**
 * Created by Konstantin on 26.05.2017.
 */

public class YandexToDomainConverter {

    public static List<TrainThread> scheduleResponseToDomain(ScheduleResponse trainScheduleResponse) {
        List<TrainThread> result = new ArrayList<>();
        if (trainScheduleResponse != null && trainScheduleResponse.getThreads() != null && trainScheduleResponse.getThreads().size() > 0) {
            for (ScheduleResponse.TrainThread yandexThread : trainScheduleResponse.getThreads()) {
                if (yandexThread.getDeparture().before(new Date())) {
                    continue;
                }
                TrainThread domainThread = new TrainThread();
                domainThread.setArrival(yandexThread.getArrival());
                domainThread.setDeparture(yandexThread.getDeparture());
                domainThread.setTitle(yandexThread.getThread().getShortTitle());

                if (yandexThread.getThread().getShortTitle() != null) {
                    String[] stations = yandexThread.getThread().getShortTitle().split(" - ");
                    domainThread.setFrom(stations[0]);
                    domainThread.setTo(stations[1]);
                }
                result.add(domainThread);
            }
            Collections.sort(result, new Comparator<TrainThread>() {
                @Override
                public int compare(TrainThread lhs, TrainThread rhs) {
                    return lhs.getDeparture().compareTo(rhs.getDeparture());
                }
            });
        }
        return result;
    }
}

File: widget/src/main/java/ru/kest/trainswidget/converters/YandexToDomainConverter.kt
package ru.kest.trainswidget.converters

import ru.kest.trainswidget.model.domain.TrainComfortLevel
import ru.kest.trainswidget.model.domain.TrainThread
import ru.kest.trainswidget.model.yandex.ScheduleResponse
import java.util.*

/**
 * DTO to Model converter
 *
 * Created by Konstantin on 26.05.2017.
 */
object YandexToDomainConverter {

    fun scheduleResponseToDomain(trainScheduleResponse: ScheduleResponse?): List<TrainThread> {
        val result = ArrayList<TrainThread>()
        if (trainScheduleResponse != null && trainScheduleResponse.threads.isNotEmpty()) {
            for ((arrival, departure, _, thread) in trainScheduleResponse.threads) {
                if (departure!!.before(Date())) {
                    continue
                }

                val domainThread = TrainThread(arrival = arrival!!, departure = departure, title = thread?.shortTitle ?: "")
                if (thread != null) {
                    if (thread.shortTitle != null) {
                        val stations = thread.shortTitle.split(" — ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                        domainThread.from = stations[0]
                        domainThread.to = stations[1]
                    }
                    if (thread.number != null) {
                        domainThread.comfortLevel = TrainComfortLevel.getComfortLevel(thread.number)
                    }
                }
                result.add(domainThread)
            }
            Collections.sort(result) { (_, departure1), (_, departure2) -> departure1.compareTo(departure2) }
        }
        return result
    }
}
